File: 02-programacion-basica/10-validacion-errores-y-excepciones/java/soluciones/RetoEdadValida.java
import java.util.Scanner;

public class RetoEdadValida {
    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {
            while (true) {
                System.out.print("Edad: ");
                String entrada = scanner.nextLine().trim();
                try {
                    int edad = Integer.parseInt(entrada);
                    if (edad < 0 || edad > 120) {
                        System.out.println("Edad fuera de rango");
                        continue;
                    }
                    System.out.println("Edad válida: " + edad);
                    break;
                } catch (NumberFormatException e) {
                    // La excepción distingue un formato incorrecto de un valor fuera de rango.
                    System.out.println("Debe ingresar un número entero");
                }
            }
        }
    }
}

File: 02-programacion-basica/10-validacion-errores-y-excepciones/scala/soluciones/RetoEdadValida.scala
import scala.io.StdIn.readLine
import scala.util.Try

@main def retoEdadValida(): Unit =
  var valida = false
  while !valida do
    val entrada = readLine("Edad: ").trim
  // Try convierte un posible fallo de conversión en un resultado que puede evaluarse.
    Try(entrada.toInt).toOption match
      case Some(edad) if edad >= 0 && edad <= 120 =>
        println(s"Edad válida: $edad")
        valida = true
      case Some(_) => println("Edad fuera de rango")
      case None => println("Debe ingresar un número entero")
